tests/unittests/sample_tests.py

```diff
@@ -858,6 +858,24 @@ def test_vector_array_fields(self):
             self.assertIsInstance(fields["vector_field"], fo.VectorField)
             self.assertIsInstance(fields["array_field"], fo.ArrayField)
 
+    @drop_datasets
+    def test_reserved_keywords_fields(self):
+        with self.assertRaises(ValueError):
+            fo.Sample(filepath="img.png", pk=321)
+
+        sample1 = fo.Sample(
+            filepath="img.png",
+        )
+
+        with self.assertRaises(ValueError):
+            sample1.pk = 321
+
+        with self.assertRaises(ValueError):
+            sample1["pk"] = 321
```

Comment:
It's been awhile since I've thought about this issue, but back when I created it, I thought to myself that there's likely a way to allow `sample["xxx"]` to work for *any* possible field name without restrictions.

`sample.xxx` will never be possible in many situations, even if we prevent `sample.pk` in particular. For example, a user could never use `Sample.__getattr__` to access a field they added that happens to match any of the public method names on the `Sample` class, either.

Reply:
Internally, the mongoengine `Document` class upon which the entire FO data model is built maintains the actual values in a `_data` dict. `__getattr__` always has to compete with other public symbols of the `Document` class, but `__getitem__` can directly access the values from the `_data` dict and therefore is always a "safe" way to access arbitrarily-named fields.